src/metabase/upload.clj

```diff
@@ -612,7 +620,12 @@
           normed-name->field (m/index-by (comp normalize-column-name :name)
                                          (t2/select :model/Field :table_id (:id table) :active true))
           normed-header      (map normalize-column-name header)
-          create-auto-pk?    (not (contains? normed-name->field auto-pk-column-name))
+          ;; Create auto-pk columns for drivers that supported uploads before auto-pk columns
+          ;; were introduced by metabase#36249. Otherwise we can assume that the table was created
+          ;; with an auto-pk column, or it has been deliberately removed.
+          create-auto-pk?    (and
+                              (contains? #{:mysql :postgres :h2} (:engine database))
```

Comment:
I think redshift is a special case here, I prefer writing this as negation : `(not #{:redshift} (:engine database))`.

Also, maybe it's better if we turn this into a `should-create-auto-pk?` helper.

Reply:
It is possible for 3rd party drivers to have implemented CSV uploads already, and indeed it would be a nice-to-have to have them create an auto PK here. Creating a `should-create-auto-pk?` driver multimethod would solve the problem, but I considered it not worth it. It's unlikely that a 3rd party driver author cares enough about creating an auto PK column to implement the method (we're only doing this as a nice-to-have to anticipate future actions features).

Reply:
I decided to create this as a multimethod called `driver/create-auto-pk-with-append-csv?` in 6a6a2e00bab0caf80144d8b796fd0c75b9ada2a9. The thing that tipped me over the edge was this conversation with Chris https://github.com/metabase/metabase/pull/37505#discussion_r1473562804